framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -381,9 +378,80 @@ public void processClassTree(ClassTree classTree) {
 
         checkExtendsImplements(classTree);
 
+        checkPolymorphicClass(classTree);
+
         super.visitClass(classTree, null);
     }
 
+    /**
+     * Throws an error if {@code classTree} has polymorphic fields but is not annotated with
+     * {@code @HasQualifierParameter}.
+     */
+    private void checkPolymorphicClass(ClassTree classTree) {
+        List<? extends Tree> members = classTree.getMembers();
+        Set<? extends AnnotationMirror> topAnnotations =
+                atypeFactory.getQualifierHierarchy().getTopAnnotations();
+
+        for (Tree mem : members) {
+            if (mem.getKind() == Tree.Kind.VARIABLE) {
+                AnnotatedTypeMirror fieldAnno = atypeFactory.getAnnotatedType(mem);
+                for (AnnotationMirror top : topAnnotations) {
+                    AnnotationMirror polyAnnoInHierarchy =
+                            atypeFactory.getQualifierHierarchy().getPolymorphicAnnotation(top);
+                    if (polyAnnoInHierarchy != null) {
+                        PolyTypeScanner polyScanner = new PolyTypeScanner();
+                        if (polyScanner.visit(fieldAnno)) {
+                            Element classTreeElement = TreeUtils.elementFromTree(classTree);
+                            if (!atypeFactory.hasQualifierParameter(classTreeElement)) {
+                                checker.report(
+                                        Result.failure("invalid.polymorphic.qualifier.use"), mem);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /** A scanner that indicates whether any (sub-)types are annotated as polymorphic. */
+    class PolyTypeScanner extends SimpleAnnotatedTypeScanner<Boolean, Void> {
```

Comment:
If a nested class doesn't require access to an instance of the outer class, then it should be marked `static`.

Reply:
I am accessing atypeFactory in the defaultActionMethod.